File: src/IsbInterop/Base/Wrappers/IsbObject.vb
﻿Imports IsbInterop.Data
Imports IsbInterop.DataTypes.Enumerable
Imports IsbInterop.Presentation
Imports IsbInterop.Presentation.Wrappers

Namespace Base.Wrappers

  ''' <summary>
  ''' Обертка над IObject.
  ''' </summary>
  Friend MustInherit Class IsbObject(Of TI As IObjectInfo)
    Inherits BaseIsbObject
    Implements IObject(Of TI)

#Region "Поля и свойства"

    ''' <summary>
    ''' Признак открытости набора данных.
    ''' </summary>
    Public ReadOnly Property Active As Boolean Implements IObject(Of TI).Active
      Get
        Return GetRcwProperty("Active")
      End Get
    End Property

    ''' <summary>
    ''' Тип объекта.
    ''' </summary>
    Public ReadOnly Property ComponentType As TCompType Implements IObject(Of TI).ComponentType
      Get
        Return GetRcwProperty("ComponentType")
      End Get
    End Property

    ''' <summary>
    ''' ИД.
    ''' </summary>
    Public ReadOnly Property Id As Integer Implements IObject(Of TI).Id
      Get
        Return GetRcwProperty("ID")
      End Get
    End Property

    ''' <summary>
    ''' Имя объекта.
    ''' </summary>
    Public ReadOnly Property Name As String Implements IObject(Of TI).Name
      Get
        Return DirectCast(GetRcwProperty("Name"), String)
      End Get
    End Property

    ''' <summary>
    ''' Имя таблицы в базе данных.
    ''' </summary>
    Public ReadOnly Property SqlTableName As String Implements IObject(Of TI).SqlTableName
      Get
        Return DirectCast(GetRcwProperty("SQLTableName"), String)
      End Get
    End Property

    ''' <summary>
    ''' Состояние набора данных.
    ''' </summary>
    Public ReadOnly Property State As TDataSetState Implements IObject(Of TI).State
      Get
        Return GetRcwProperty("State")
      End Get
    End Property

#End Region

#Region "Методы"

    ''' <summary>
    ''' Добавляет условие ограничения в запрос набора данных.
    ''' </summary>
    ''' <param name="queryWhereSection">Секция where запроса.</param>
    ''' <returns>ИД условия в запросе.</returns>
    Public Function AddWhere(queryWhereSection As String) As Integer Implements IObject(Of TI).AddWhere
      Return InvokeRcwInstanceMethod("AddWhere", queryWhereSection)
    End Function

    ''' <summary>
    ''' Удаляет условие ограничения набора данных.
    ''' </summary>
    ''' <param name="queryConditionId">ИД условия в запросе.</param>
    Public Sub DelWhere(queryConditionId As Integer) Implements IObject(Of TI).DelWhere
      InvokeRcwInstanceMethod("DelWhere", queryConditionId)
    End Sub

    ''' <summary>
    ''' Завершить работу с объектом.
    ''' </summary>
    Public Sub DoFinalize() Implements IObject(Of TI).DoFinalize
      ' Для записей справочников, записей DataSet: если внутри using изменили запись и свалился наш код с исключением,
      ' то Finalize пытается закрыть запись, но закрыть несохраненную запись нельзя.
      ' Клиентский код должен корректно обрабатывать такие ситуации.
      InvokeRcwInstanceMethod("Finalize")
    End Sub

    ''' <summary>
    ''' Получает детальный раздел набора данных.
    ''' </summary>
    ''' <param name="dataSetNumber">Номер детального раздела.</param>
    ''' <returns>Детальный раздел.</returns>
    Public Function GetDetailDataSet(dataSetNumber As Integer) As IDataSet Implements IObject(Of TI).GetDetailDataSet
      Dim rcwDataSet = InvokeRcwInstanceMethod("DetailDataSet", dataSetNumber)

      Return New Data.Wrappers.DataSet(rcwDataSet, Scope)
    End Function

    ''' <summary>
    ''' Получает окружение.
    ''' </summary>
    ''' <typeparam name="TP">Тип параметров.</typeparam>
    ''' <returns>Список переменных окружения объекта.</returns>
    Public Function GetEnvironment(Of TP As IBaseIsbObject)() As Accessory.IList(Of TP) Implements IObject(Of TI).GetEnvironment
      Dim rcwEnvironment = GetRcwProperty("Environment")
      Return New Accessory.Wrappers.List(Of TP)(rcwEnvironment, Scope)
    End Function

    ''' <summary>
    ''' Получает форму-карточку текущего представления объекта.
    ''' </summary>
    ''' <returns>Форма-карточка.</returns>
    Public Function GetForm() As IForm Implements IObject(Of TI).GetForm
      Dim rcwForm = GetRcwProperty("Form")
      Return New Form(rcwForm, Scope)
    End Function

    ''' <summary>
    ''' Получает параметры объекта.
    ''' </summary>
    ''' <typeparam name="TP">Тип параметров.</typeparam>
    ''' <returns>Список параметров объекта.</returns>
    Public Function GetParams(Of TP As IBaseIsbObject)() As Accessory.IList(Of TP) Implements IObject(Of TI).GetParams
      Dim rcwParams = GetRcwProperty("Params")
      Return New Accessory.Wrappers.List(Of TP)(rcwParams, Scope)
    End Function

    ''' <summary>
    ''' Получает информацию об объекте.
    ''' </summary>
    ''' <returns>Информация об объекте.</returns>
    Public MustOverride Function GetInfo() As TI Implements IObject(Of TI).GetInfo

    ''' <summary>
    ''' Получает реквизит набора данных.
    ''' </summary>
    ''' <param name="requisiteName">Имя реквизита.</param>
    ''' <returns>Реквизит.</returns>
    ''' <remarks>Умышленно делаем метод виртуальным, чтобы использовать вместе с IRequisiteAutoCleaner.</remarks>
    Public Function GetRequisite(requisiteName As String) As IRequisite Implements IObject(Of TI).GetRequisite
      Dim rcwRequisite = GetRcwProperty("Requisites", requisiteName)
      Return DerivedRequisiteFactory.CreateRequisite(rcwRequisite, Scope)
    End Function

    ''' <summary>
    ''' Обновляет набор данных.
    ''' </summary>
    Public Sub Refresh() Implements IObject(Of TI).Refresh
      InvokeRcwInstanceMethod("Refresh")
    End Sub

    ''' <summary>
    ''' Сохраняет объект.
    ''' </summary>
    Public Sub Save() Implements IObject(Of TI).Save
      InvokeRcwInstanceMethod("Save")
    End Sub

    ''' <summary>
    ''' Получить COM-объект IObjectInfo.
    ''' </summary>
    ''' <returns>COM-объект IObjectInfo.</returns>
    Protected Function GetRcwObjectInfo() As Object
      Return GetRcwProperty("Info")
    End Function

#End Region

#Region "Конструкторы"

    ''' <summary>
    ''' Конструктор.
    ''' </summary>
    ''' <param name="externalObject">Внешний объект.</param>
    ''' <param name="scope">Область видимости.</param>
    Protected Sub New(externalObject As Object, scope As IScope)
      MyBase.New(externalObject, scope)
    End Sub

#End Region

  End Class
End Namespace

File: src/IsbInterop.CSharp/Base/Wrappers/IsbObject.cs
using IsbInterop.Accessory;
using IsbInterop.Accessory.Wrappers;
using IsbInterop.Data;
using IsbInterop.Data.Wrappers;
using IsbInterop.DataTypes.Enumerable;
using IsbInterop.Presentation;
using IsbInterop.Presentation.Wrappers;

namespace IsbInterop.Base.Wrappers
{
  /// <summary>
  /// Обертка над IObject.
  /// </summary>
  internal abstract class IsbObject<TI> : BaseIsbObject, IObject<TI>
    where TI : IObjectInfo
  {
    #region IDisposable

    //private bool disposed = false;

    //// TODO ksvadim: Лучше вызывать метод Finalize явно, где это нужно (например при освобождении документа).
    //// Если показать форму-список через ComponentShow и затем сразу вызывается Finalize у IComponent, то возникает ошибка.
    ///// <summary>
    ///// Освободить объект.
    ///// </summary>
    ///// <param name="disposing">Признак, что метод Dispose вызван явно.</param>
    //protected override void Dispose(bool disposing)
    //{
    //  if (disposed)
    //    return;

    //  if (disposing)
    //  {
    //    if (RcwObject != null)
    //      DoFinalize();
    //  }

    //  disposed = true;

    //  base.Dispose(disposing);
    //}

    #endregion

    #region Поля и свойства

    /// <summary>
    /// Признак открытости набора данных.
    /// </summary>
    public bool Active => (bool)GetRcwProperty("Active");

    /// <summary>
    /// Тип объекта.
    /// </summary>
    public TCompType ComponentType => (TCompType)GetRcwProperty("ComponentType");

    /// <summary>
    /// ИД.
    /// </summary>
    public int Id => (int)GetRcwProperty("ID");

    /// <summary>
    /// Имя объекта.
    /// </summary>
    public string Name => (string)GetRcwProperty("Name");

    /// <summary>
    /// Имя таблицы в базе данных.
    /// </summary>
    public string SqlTableName => (string)GetRcwProperty("SQLTableName");

    /// <summary>
    /// Состояние набора данных.
    /// </summary>
    public TDataSetState State => (TDataSetState)GetRcwProperty("State");

    #endregion

    #region Методы

    /// <summary>
    /// Добавляет условие ограничения в запрос набора данных.
    /// </summary>
    /// <param name="queryWhereSection">Секция where запроса.</param>
    /// <returns>ИД условия в запросе.</returns>
    public int AddWhere(string queryWhereSection)
    {
      return (int)InvokeRcwInstanceMethod("AddWhere", queryWhereSection);
    }

    /// <summary>
    /// Удаляет условие ограничения набора данных.
    /// </summary>
    /// <param name="queryConditionId">ИД условия в запросе.</param>
    public void DelWhere(int queryConditionId)
    {
      InvokeRcwInstanceMethod("DelWhere", queryConditionId);
    }

    /// <summary>
    /// Завершает работу с объектом.
    /// </summary>
    public void DoFinalize()
    {
      // Для записей справочников, записей DataSet: если внутри using изменили запись и свалился наш код с исключением,
      // то Finalize пытается закрыть запись, но закрыть несохраненную запись нельзя.
      // Клиентский код должен корректно обрабатывать такие ситуации.
      InvokeRcwInstanceMethod("Finalize");
    }

    /// <summary>
    /// Получает детальный раздел набора данных.
    /// </summary>
    /// <param name="dataSetNumber">Номер детального раздела.</param>
    /// <returns>Детальный раздел.</returns>
    public IDataSet GetDetailDataSet(int dataSetNumber)
    {
      var rcwDataSet = InvokeRcwInstanceMethod("DetailDataSet", dataSetNumber);
      return new DataSet(rcwDataSet, Scope);
    }

    /// <summary>
    /// Получает окружение.
    /// </summary>
    /// <typeparam name="TP">Тип параметров.</typeparam>
    /// <returns>Список переменных окружения объекта.</returns>
    public IList<TP> GetEnvironment<TP>() where TP : IBaseIsbObject
    {
      var rcwEnvironment = GetRcwProperty("Environment");
      return new List<TP>(rcwEnvironment, Scope);
    }

    /// <summary>
    /// Получает форму-карточку текущего представления объекта.
    /// </summary>
    /// <returns>Форма-карточка.</returns>
    public IForm GetForm()
    {
      var rcwForm = GetRcwProperty("Form");
      return new Form(rcwForm, Scope);
    }

    /// <summary>
    /// Получает параметры объекта.
    /// </summary>
    /// <typeparam name="TP">Тип параметра.</typeparam>
    /// <returns>Список параметров объекта.</returns>
    public IList<TP> GetParams<TP>() where TP : IBaseIsbObject
    {
      var rcwParams = GetRcwProperty("Params");
      return new List<TP>(rcwParams, Scope);
    }

    /// <summary>
    /// Получает информацию об объекте.
    /// </summary>
    /// <returns>Информация об объекте.</returns>
    public abstract TI GetInfo();

    /// <summary>
    /// Получает реквизит.
    /// </summary>
    /// <param name="requisiteName">Имя реквизита.</param>
    /// <returns>Реквизит.</returns>
    /// <remarks>Умышленно делаем метод виртуальным, чтобы использовать вместе с IRequisiteAutoCleaner.</remarks>
    public IRequisite GetRequisite(string requisiteName)
    {
      var rcwRequisite = GetRcwProperty("Requisites", requisiteName);
      return DerivedRequisiteFactory.CreateRequisite(rcwRequisite, Scope);
    }

    /// <summary>
    /// Обновить набор данных.
    /// </summary>
    public void Refresh()
    {
      InvokeRcwInstanceMethod("Refresh");
    }

    /// <summary>
    /// Сохраняет объект.
    /// </summary>
    public void Save()
    {
      InvokeRcwInstanceMethod("Save");
    }

    /// <summary>
    /// Получить COM-объект IObjectInfo.
    /// </summary>
    /// <returns>COM-объект IObjectInfo.</returns>
    protected object GetRcwObjectInfo()
    {
      return GetRcwProperty("Info");
    }

    #endregion

    #region Конструкторы

    /// <summary>
    /// Конструктор.
    /// </summary>
    /// <param name="externalObject">Внешний объект.</param>
    /// <param name="scope">Область видимости.</param>
    protected IsbObject(object externalObject, IScope scope) : base(externalObject, scope) { }

    #endregion
  }
}
